Add failure only unit test reporter

bachase opened the pull request:
Adds a new unit test reporter that only reports failures to the console. It is enabled by adding `-q/--quiet` to  the command line when invoking unit tests.  It does not report `log` test output.

Example output
```
$ ./RippleD.exe -q -u
Loading: "testSetup14\rippled.cfg"
Loading: "testSetup14\rippled.cfg"
Loading: "testSetup15\rippled.cfg"
Loading: "testSetup15\rippled.cfg"
ripple.protocol.Quality comparisons #1 failed: Quality_test.cpp(278)
ripple.protocol.Quality round #1 failed: Quality_test.cpp(247)
ripple.protocol.Quality round #2 failed: Quality_test.cpp(248)
134 suites, 632 cases, 326391 tests total, 3 failures
```

Addresses VinnieFalco/Beast#166.

vinniefalco commented:
**LOVE IT!!!**

bachase commented:
About to force push a fix . . . I forget to update vcxproj

nbougalis commented:
Not a reviewer, but 👍 on this! 💯

vinniefalco commented:
This is exactly what we need, instead of fiddling with the predefined reporter, we just make a new one that does what we want. And if we don't like it we can always switch. So rippled can customize the unit test output to its heart's desire without the need for architectural changes in the unit test framework. Great!
:+1:

wilsonianb commented:
Should we use this with CI tests, or do we want to keep those verbose?

vinniefalco commented:
On CI I like seeing only the failures and totals.

bachase commented:
Added the CI change but can pull it if anyone objects.

ximinez commented:
Will we lose information in CI if we have a rare, hard to reproduce, but non-spurious failure condition? I'm specifically thinking of `log`, here...

mellery451 commented on src/test/failure_reporter.h:
```diff
@@ -0,0 +1,171 @@
+
+#include <beast/unit_test/amount.hpp>
+#include <beast/unit_test/recorder.hpp>
+#include <boost/lexical_cast.hpp>
```
I couldn't see any obvious reason to include the following:

```
#include <boost/lexical_cast.hpp>
#include <boost/optional.hpp>
#include <algorithm>
#include <chrono>
#include <functional>
#include <sstream>
#include <iomanip>
```
...so I did a nonunity build with these removed and it worked fine. You might consider removing them.

bachase replied:
Good catch. I'll fix.

bachase commented:
That is a good point @ximinez. I dropped the log output because there is a good bit of it now but leaving it in CI might be prudent.

ximinez commented:
I have a suggestion: https://github.com/ximinez/rippled/commit/63f02a21ad4ec917bd400fca7e9ad5ca9112de9a

ximinez commented:
Is it possible to get the "Longest suite times" and time summary back?

bachase commented:
Squashed and added @ximinez's commit.

* Running with `-q|--quiet` now prints the timing summary.
* Running with `-q|--quiet` and `--unittest-arg=log` adds the `log` output.  I used this for CI to be sure we can catch rare issues as @ximinez pointed out.

Quiet mode
```
$ ./RippleD.exe -u -q
Loading: "testSetup14\rippled.cfg"
Loading: "testSetup14\rippled.cfg"
Loading: "testSetup15\rippled.cfg"
Loading: "testSetup15\rippled.cfg"
Longest suite times:
   10.3s ripple.app.Flow
    5.9s ripple.app.RPCOverload
    5.9s ripple.server.ServerStatus
    5.6s ripple.tx.Offer
    5.3s ripple.app.Path
    4.8s ripple.app.Book
    4.2s ripple.app.LedgerData
    3.4s ripple.app.TxQ
    2.5s ripple.app.LedgerRequestRPC
    2.4s ripple.app.SusPay
87.8s, 134 suites, 632 cases, 326376 tests total, 0 failures
```

Quiet + log mode for CI
```
$./RippleD.exe -u -q --unittest-arg=log
beast.utility.Debug _DEBUG              = 0
BEAST_DEBUG         = 0
BEAST_FORCE_DEBUG   = 0
sizeof(std::size_t) = 8
ripple.app.Flow Line Quality FTL:Application Application::onStop took 52ms
ripple.app.LedgerLoad Load ledger: Bad Files ERR:Application The specified ledger could not be loaded.
ripple.app.LedgerLoad Load ledger: Bad Files FTL:Application Unable to open file 'badfile.json'
ripple.app.LedgerLoad Load ledger: Bad Files ERR:Application The specified ledger could not be loaded.
ripple.app.LedgerLoad Load ledger: Bad Files FTL:Application Unable to parse ledger JSON
ripple.app.LedgerLoad Load ledger: Bad Files ERR:Application The specified ledger could not be loaded.
ripple.app.LedgerLoad Load by index FTL:Application Application::onStop took 11ms
....
....
Longest suite times:
    8.1s ripple.app.Flow
    6.1s ripple.server.ServerStatus
    5.8s ripple.app.RPCOverload
    5.4s ripple.tx.Offer
    4.3s ripple.app.Path
    3.4s ripple.app.Book
    3.3s ripple.app.TxQ
    2.9s ripple.app.LedgerData
    2.7s ripple.app.Env
    2.3s ripple.protocol.SecretKey
81.1s, 134 suites, 632 cases, 326482 tests total, 0 failures
```

vinniefalco commented:
`--unittest-arg` is for when you want to pass options to a particular unit test. If you want to have a "log" type switch then it should just be a regular switch (e,g, `--unittest-log`). And you should set it by calling a non-const member on the reporter.

bachase commented:
Squashed and rebased on 0.60.0-b5

seelabs commented:
In 0.60.0-b7
